File: C++/19. Remove Nth Node From End of List.cpp
class Solution {
public:
    int getLength(ListNode *head)
    {
        int len = 0;
        ListNode *temp = head;
        while (temp != NULL)
        {
            temp = temp->next;
            len++;
        }
        return len;
    }
    ListNode* removeNthFromEnd(ListNode*& head, int n) {
        int len = getLength(head);
        if (n == len)
        {
            ListNode *temp = head;
            head = head->next;
            temp->next = NULL;
            delete temp;
            return head;
        }
        ListNode *prev = head;
        for (int i = 1; i < len - n; i++)
        {
            prev = prev->next;
        }
        ListNode *temp = prev->next;
        prev->next = prev->next->next;
        temp->next = NULL;
        delete temp;
        return head;
    }
};

File: JAVA/19. Remove Nth Node From End of List.java
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode sentinel = new ListNode(-1, head);
        ListNode fastPointer = sentinel;
        ListNode slowPointer = null;
        while(n > 0){
            fastPointer = fastPointer.next;
            n--;
        }
        slowPointer = sentinel;
        while(fastPointer.next != null){
            fastPointer = fastPointer.next;
            slowPointer = slowPointer.next;
        }
        slowPointer.next = slowPointer.next.next;
        return sentinel.next;

    }
}
